README.md
# base_scaffold

[![pub package](https://img.shields.io/pub/v/base_scaffold.svg)](https://pub.dev/packages/base_scaffold)

A Flutter package for base **scaffold**. After importing this plugin to your project as usual, import base_scaffold.dart and use the BaseScaffold() for ready feature like toolbar, loaderView and etc...

## Demo [(example/base_scaffold_demo)](https://github.com/Pinkal7600/base_scaffold/tree/master/example)
<img src="https://raw.githubusercontent.com/Pinkal7600/base_scaffold/master/example/demo.gif" height="400" />

## Getting Started

After importing this plugin to your project as usual, import below to
use base_scaffold.

```dart
import 'package:base_scaffold/base_scaffold.dart';  
```

##### You are ready to use the base_scaffold.

### How to use?

```dart
@override
  Widget build(BuildContext context) {
    return BaseScaffold(
      body: myBody(),

      toolbarTitle: 'Team',
      toolbarRightText: 'Add',
      toolbarRightTextClick: () {
        // your event
      },

      bodyBackgroundColor: Colors.white,
    );
  }
```

## Properties:
These are below available properties for base_scaffold.

| Properties | type | details |
|--------|-----|-----|
| body | Widget | It contains the body of the screen which will be display in safe area and between toolbar and bottom navigation bar. |
| showToolbar | bool | Show/Hide toolbar. Default is true |
| toolbar | Widget | To design custom toolbar. |
| toolbarLeftIcon | dynamic | Set icon or image on left side of toolbar. Icon can be String for Image.assets or Icon, it will be depend on `toolbarLeftIconType`. |
| toolbarLeftIconType | String | Type of left Icon in toolbar. It can be image assets or Icon. For e.g : AppToolbar.IMAGE_TYPE_IMAGE or AppToolbar.IMAGE_TYPE_ICON. |
| toolbarLeftIconClick | Function | Get callback click of left icon. |
| toolbarRightIcon | dynamic | Set icon or image on right side of toolbar. Icon can be String for Image.assets or Icon, it will be depend on `toolbarRightIconType`. |
| toolbarRightIconType | String | Type of right Icon in toolbar. It can be image assets or Icon. For e.g : AppToolbar.IMAGE_TYPE_IMAGE or AppToolbar.IMAGE_TYPE_ICON. |
| toolbarRightIconClick | Function | Get callback click of rith icon. |
| toolbarTitle | String | Title of toolbar. |
| toolbarRightText | String | Set some text to right instead of icon. |
| toolbarRightTextClick | Function | Get callback click of right text. |
| toolbarLeftText | String | Set some text to left instead of icon. |
| toolbarLeftTextClick | Function | Get callback click of left text. |
| toolbarBackgroundColor | Color | Set toolbar background color. |
| toolbarHeight | double | Set custom toolbar height as per requirement. Minimum height should be 50. |
| toolbarLeftTextStyle | TextStyle | Set custom text style for left toolbar text. |
| toolbarTitleTextStyle | TextStyle | Set custom text style for title toolbar text. |
| toolbarRightTextStyle | TextStyle | Set custom text style for right toolbar text. |
| isScreenLoading | bool | Set bool to show/hide loader view or entire screen body. |
| isScreenLoadingWithBackground | bool | Set bool to show/hide loader view on screen body. |
| screenLoaderView | Widget | Set custom widget for in screen loader view. |
| backgroundScreenLoaderView | Widget | Set custom widget for on screen loader view. |
| scaffoldBackgroundColor | Color | Set scaffold background color. |
| bodyBackgroundColor | Color | Set body background color. |

**Note** : For more base_scaffold properties and details please check here : https://pub.dev/documentation/base_scaffold/latest/base_scaffold/BaseScaffold-class.html


CHANGELOG.md
## 1.2.1

* Null safety

## 1.2.0

* Flutter 2.0 support

## 1.1.5

* Bug fixed for New Flutter 1.17.0

## 1.1.4

* Bug fixed.

## 1.1.3

* Added comment on all properties in base_scaffold and base_toolbar.

## 1.1.2

* Minor fixes.

## 1.1.1

* Support all scaffold properties in base_scaffold.

## 1.1.0

* Readme update.

## 1.0.9

* Issue fixed.

## 1.0.8

* Set custom widget child to left, right and center of toolbar.

## 1.0.7

* Minor Changes.

## 1.0.6

* Added readme in example demo.

## 1.0.5

* URL fix to Demo gif in readme.

## 1.0.4

* Demo gif added to readme.

## 1.0.3

* Added new example with use case of toolbar, screen loader and on screen loader.
* Resolve issue regarding null body.
* Added TextStyle for toolbar related text.

## 1.0.2

* Rename file from app_toolbar to base_toolbar.

## 1.0.1

* Change description.

## 1.0.0

* Added description for this package use and information how to use with example.

## 0.0.2

* Added CHANGELOG.md.

## 0.0.1

* Initial release.

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';


void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {


    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/base_scaffold_demo/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:base_scaffold_demo/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/app_toolbar.dart
import 'package:flutter/material.dart';

class AppToolbar extends StatefulWidget {
  static const String IMAGE_TYPE_IMAGE = "image";
  static const String IMAGE_TYPE_ICON = "icon";

  final leftIcon;
  final String leftIconType;
  final Function leftIconClick;

  final rightIcon;
  final String rightIconType;
  final Function rightIconClick;

  final String title;
  final String rightText;
  final Function rightTextClick;
  final String leftText;
  final Function leftTextClick;

  final Color backgroundColor;

  AppToolbar({
    this.leftIconType,
    this.leftIcon,
    this.leftIconClick,
    this.rightIcon,
    this.rightIconType,
    this.rightIconClick,
    this.title,
    this.rightText,
    this.rightTextClick,
    this.leftText,
    this.leftTextClick,
    this.backgroundColor = Colors.white,
  });

  @override
  _AppToolbarState createState() => _AppToolbarState();
}

class _AppToolbarState extends State<AppToolbar> {
  ThemeData themeData;

  @override
  Widget build(BuildContext context) {
    themeData = Theme.of(context);
    return Container(
      color: widget.backgroundColor,
      height: 60,
      child: Row(
        children: <Widget>[
          Expanded(
              flex: 1,
              child: widget.leftIcon != null
                  ? leftIconWidget(widget.leftIconClick)
                  : leftTextTitle(widget.leftText, widget.leftTextClick)),
          Expanded(flex: 4, child: centerText(widget.title)),
          Expanded(
              flex: 1,
              child: widget.rightIcon != null
                  ? rightIconWidget(widget.rightIconClick)
                  : rightTextTitle(widget.rightText, widget.rightTextClick)),
//        leftIconWidget(leftIconClick),
        ],
      ),
    );
  }

  Widget leftIconWidget(leftIconClick) {
    return GestureDetector(
      onTap: leftIconClick,
      child: Padding(
        padding: EdgeInsets.all(15),
        child: getLeftIcon(widget.leftIconType, widget.leftIcon),
      ),
    );
  }

  Widget getLeftIcon(leftIconType, leftIcon) {
    if (leftIconType == AppToolbar.IMAGE_TYPE_IMAGE) {
      return Image.asset(
        leftIcon,
        height: 30,
        width: 30,
      );
    } else if (leftIconType == AppToolbar.IMAGE_TYPE_ICON) {
      return Icon(
        leftIcon,
        color: themeData.primaryColor,
      );
    } else {
      return Container();
    }
  }

  Widget leftTextTitle(String leftText, leftTextClick) {
    return Container(
      child: leftText == null || leftText.isEmpty
          ? Container()
          : GestureDetector(
        onTap: leftTextClick,
        child: Text(
          leftText,
          style: themeData.textTheme.subtitle
              .copyWith(color: themeData.primaryColor),
          textAlign: TextAlign.center,
        ),
      ),
    );
  }

  Widget centerText(String title) {
    return Container(
      child: title == null || title.isEmpty
          ? Container()
          : Text(
        title,
        style: themeData.textTheme.title
            .copyWith(color: Colors.black, fontWeight: FontWeight.w600),
        textAlign: TextAlign.center,
      ),
    );
  }

  Widget rightTextTitle(String rightText, rightTextClick) {
    return Container(
      child: rightText == null || rightText.isEmpty
          ? Container()
          : GestureDetector(
        onTap: rightTextClick,
        child: Text(
          rightText,
          style: themeData.textTheme.subtitle
              .copyWith(color: themeData.primaryColor),
          textAlign: TextAlign.center,
        ),
      ),
    );
  }

  Widget rightIconWidget(rightIconClick) {
    return GestureDetector(
      onTap: rightIconClick,
      child: Padding(
        padding: EdgeInsets.all(15),
        child: getRightIcon(widget.rightIconType, widget.rightIcon),
      ),
    );
  }

  Widget getRightIcon(rightIconType, rightIcon) {
    if (rightIconType == AppToolbar.IMAGE_TYPE_IMAGE) {
      return Image.asset(
        rightIcon,
        height: 30,
        width: 30,
      );
    } else if (rightIconType == AppToolbar.IMAGE_TYPE_ICON) {
      return Icon(
        rightIcon,
        color: themeData.primaryColor,
      );
    } else {
      return Container();
    }
  }
}


lib/base_scaffold.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import 'base_toolbar.dart';

class BaseScaffold extends StatefulWidget {
  /// It contains the body of the screen which will be display in safe area
  /// and between toolbar and bottom navigation bar.
  final Widget? body;

  /// Design your own custom toolbar.
  final Widget? toolbar;

  /// Design custom widget for loader view where body will display empty.
  final Widget? screenLoaderView;

  /// Design custom widget for loader view which will display with
  /// body content.
  final Widget? backgroundScreenLoaderView;

  /// Set true to show loading.
  /// @Note: This will hide the body and display empty screen.
  final bool isScreenLoading;

  /// Set true to show loading with body in background.
  /// @Note: This will show body in background and loader on it.
  final bool isScreenLoadingWithBackground;

  /// Show/Hide toolbar by true/false value.
  /// By default it's value is true.
  final bool showToolbar;

  /// Set scaffold background color.
  final Color scaffoldBackgroundColor;

  /// Set toolbar background color.
  final Color toolbarBackgroundColor;

  /// Set body background color.
  final Color bodyBackgroundColor;

  // Toolbar
  /// Set icon on left side of the toolbar.
  /// Icon can be an [String] value if you use an assets
  ///                   OR
  /// Icon can be an [IconData] if you use Icons.
  final toolbarLeftIcon;

  /// If you set left icon you need to provide the type of icon.
  /// Use this to set type of icon can be from below two option.
  /// * [BaseToolbar.IMAGE_TYPE_IMAGE] if icon is from assets.
  /// * [BaseToolbar.IMAGE_TYPE_ICON] if icon is type of Icons.
  ///
  /// @Note: If you not provide any type to left icon then left icon will
  /// not display in toolbar.
  final String? toolbarLeftIconType;

  /// Set left icon click event if you have used [toolbarLeftIcon] &
  /// [toolbarLeftIconType].
  final Function? toolbarLeftIconClick;

  /// Set icon on right side of the toolbar.
  /// Icon can be an [String] value if you use an assets
  ///                   OR
  /// Icon can be an [IconData] if you use Icons.
  final toolbarRightIcon;

  /// If you set right icon you need to provide the type of icon.
  /// Use this to set type of icon can be from below two option.
  /// * [BaseToolbar.IMAGE_TYPE_IMAGE] if icon is from assets.
  /// * [BaseToolbar.IMAGE_TYPE_ICON] if icon is type of Icons.
  ///
  /// @Note: If you not provide any type to right icon then right icon will
  /// not display in toolbar.
  final String? toolbarRightIconType;

  /// Set right icon click event if you have used [toolbarRightIcon] &
  /// [toolbarRightIconType].
  final Function? toolbarRightIconClick;

  /// Set title on toolbar.
  final String? toolbarTitle;

  /// Set text at right side of toolbar.
  final String? toolbarRightText;

  /// Set click event of right side text.
  /// @Note: It will not work if [toolbarRightText] is empty or not set.
  final Function? toolbarRightTextClick;

  /// Set text at left side of toolbar.
  final String? toolbarLeftText;

  /// Set click event of left side text.
  /// @Note: It will not work if [toolbarLeftText] is empty or not set.
  final Function? toolbarLeftTextClick;

  /// Manage height of toolbar.
  /// By default it height is [60] and it can not be less than [50].
  final double toolbarHeight;

  /// Set custom text style of left text in toolbar.
  final TextStyle? toolbarLeftTextStyle;

  /// Set custom text style of title text in toolbar.
  final TextStyle? toolbarTitleTextStyle;

  /// Set custom text style of right text in toolbar.
  final TextStyle? toolbarRightTextStyle;

  /// Use to set only color of left icon.
  final Color? toolbarLeftIconColor;

  /// Use to set only color of right icon.
  final Color? toolbarRightIconColor;

  /// Make custom widget to set at left side of toolbar.
  final Widget? toolbarLeftChild;

  /// Make custom widget to set at center of toolbar.
  final Widget? toolbarCenterChild;

  /// Make custom widget to set at right side of toolbar.
  final Widget? toolbarRightChild;

  // Scaffold
  /// Default properties of [Scaffold]
  final bool extendBody;

  /// Default properties of [Scaffold]
  final bool extendBodyBehindAppBar;

  /// Default properties of [Scaffold]
  final Widget? floatingActionButton;

  /// Default properties of [Scaffold]
  final FloatingActionButtonLocation? floatingActionButtonLocation;

  /// Default properties of [Scaffold]
  final FloatingActionButtonAnimator? floatingActionButtonAnimator;

  /// Default properties of [Scaffold]
  final List<Widget>? persistentFooterButtons;

  /// Default properties of [Scaffold]
  final Widget? drawer;

  /// Default properties of [Scaffold]
  final Widget? endDrawer;

  /// Default properties of [Scaffold]
  final Color? drawerScrimColor;

  /// Default properties of [Scaffold]
  final Widget? bottomNavigationBar;

  /// Default properties of [Scaffold]
  final Widget? bottomSheet;

  /// Default properties of [Scaffold]
  final bool? resizeToAvoidBottomInset;

  /// Default properties of [Scaffold]
  final bool primary;

  /// Default properties of [Scaffold]
  final DragStartBehavior drawerDragStartBehavior;

  /// Default properties of [Scaffold]
  final double? drawerEdgeDragWidth;

  /// Default properties of [Scaffold]

  BaseScaffold({
    this.body,
    this.toolbar,
    this.isScreenLoading = false,
    this.isScreenLoadingWithBackground = false,
    this.showToolbar = true,
    this.toolbarLeftIcon,
    this.toolbarLeftIconColor,
    this.toolbarLeftIconType,
    this.toolbarLeftIconClick,
    this.toolbarRightIcon,
    this.toolbarRightIconColor,
    this.toolbarRightIconType,
    this.toolbarRightIconClick,
    this.toolbarTitle,
    this.toolbarRightText,
    this.toolbarRightTextClick,
    this.toolbarLeftText,
    this.toolbarLeftTextClick,
    this.toolbarHeight = 60,
    this.toolbarLeftChild,
    this.toolbarCenterChild,
    this.toolbarRightChild,
    this.screenLoaderView,
    this.backgroundScreenLoaderView,
    this.bodyBackgroundColor = Colors.white,
    this.scaffoldBackgroundColor = Colors.white,
    this.toolbarBackgroundColor = Colors.white,
    this.toolbarLeftTextStyle,
    this.toolbarTitleTextStyle,
    this.toolbarRightTextStyle,
    this.floatingActionButton,
    this.floatingActionButtonLocation,
    this.floatingActionButtonAnimator,
    this.persistentFooterButtons,
    this.drawer,
    this.endDrawer,
    this.bottomNavigationBar,
    this.bottomSheet,
    this.resizeToAvoidBottomInset,
    this.primary = true,
    this.drawerDragStartBehavior = DragStartBehavior.start,
    this.extendBody = false,
    this.extendBodyBehindAppBar = false,
    this.drawerScrimColor,
    this.drawerEdgeDragWidth,
  }) : assert(toolbarHeight > 50);

  @override
  _BaseScaffoldState createState() => _BaseScaffoldState();
}

class _BaseScaffoldState extends State<BaseScaffold> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: widget.scaffoldBackgroundColor,
      body: SafeArea(
        child: scaffoldBody(),
      ),
      floatingActionButton: widget.floatingActionButton,
      floatingActionButtonLocation: widget.floatingActionButtonLocation,
      floatingActionButtonAnimator: widget.floatingActionButtonAnimator,
      persistentFooterButtons: widget.persistentFooterButtons,
      drawer: widget.drawer,
      endDrawer: widget.endDrawer,
      bottomNavigationBar: widget.bottomNavigationBar,
      bottomSheet: widget.bottomSheet,
      resizeToAvoidBottomInset: widget.resizeToAvoidBottomInset,
      primary: widget.primary,
      drawerDragStartBehavior: widget.drawerDragStartBehavior,
      extendBody: widget.extendBody,
      extendBodyBehindAppBar: widget.extendBodyBehindAppBar,
      drawerScrimColor: widget.drawerScrimColor,
      drawerEdgeDragWidth: widget.drawerEdgeDragWidth,
    );
  }

  Column scaffoldBody() {
    return Column(
      children: <Widget>[
        toolbar(),
        Expanded(
          child: Container(
            color: widget.bodyBackgroundColor,
            child: widget.isScreenLoading
                ? loadingView()
                : Stack(
                    children: <Widget>[
                      widget.body ?? Container(),
                      widget.isScreenLoadingWithBackground
                          ? loadingViewWithBackground()
                          : Container(),
                    ],
                  ),
          ),
        ),
      ],
    );
  }

  Widget toolbar() => widget.toolbar != null
      ? widget.toolbar!
      : widget.showToolbar
          ? BaseToolbar(
              leftIconType: widget.toolbarLeftIconType,
              leftIcon: widget.toolbarLeftIcon,
              leftIconClick: widget.toolbarLeftIconClick,
              rightIcon: widget.toolbarRightIcon,
              rightIconType: widget.toolbarRightIconType,
              rightIconClick: widget.toolbarRightIconClick,
              title: widget.toolbarTitle,
              rightText: widget.toolbarRightText,
              rightTextClick: widget.toolbarRightTextClick,
              leftText: widget.toolbarLeftText,
              leftTextClick: widget.toolbarLeftTextClick,
              backgroundColor: widget.toolbarBackgroundColor,
              toolbarHeight: widget.toolbarHeight,
              leftTextStyle: widget.toolbarLeftTextStyle,
              titleTextStyle: widget.toolbarTitleTextStyle,
              rightTextStyle: widget.toolbarRightTextStyle,
              leftIconColor: widget.toolbarLeftIconColor,
              rightIconColor: widget.toolbarRightIconColor,
              leftChild: widget.toolbarLeftChild,
              centerChild: widget.toolbarCenterChild,
              rightChild: widget.toolbarRightChild,
            )
          : Container();

  Widget loadingView() {
    return Center(
      child: widget.screenLoaderView != null
          ? widget.screenLoaderView
          : CupertinoActivityIndicator(
              animating: true,
              radius: 15,
            ),
    );
  }

  Widget loadingViewWithBackground() {
    return Stack(
      children: <Widget>[
        Opacity(
          opacity: 0.3,
          child: Container(
            color: Colors.black,
          ),
        ),
        Center(
          child: widget.backgroundScreenLoaderView ??
              CupertinoActivityIndicator(
                animating: true,
                radius: 15,
              ),
        ),

        /// Use below loading dialog with background
//        centerLoading(),
      ],
    );
  }

  Center centerLoading() {
    return Center(
      child: ClipRRect(
        borderRadius: BorderRadius.circular(20),
        child: Container(
          padding: EdgeInsets.symmetric(horizontal: 45, vertical: 35),
          color: Colors.white,
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              CupertinoActivityIndicator(
                animating: true,
                radius: 15,
              ),
              SizedBox(height: 10),
              Text("Loading..."),
            ],
          ),
        ),
      ),
    );
  }
}


example/base_scaffold_demo/lib/home.dart
import 'package:base_scaffold/base_scaffold.dart';
import 'package:base_scaffold/base_toolbar.dart';
import 'package:flutter/material.dart';

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  bool _isScreenLoading = false;
  bool _showLoading = false;

  @override
  void initState() {
    super.initState();
    setState(() {
      _isScreenLoading = true;
    });
    Future.delayed(Duration(seconds: 2), () {
      setState(() {
        _isScreenLoading = false;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return BaseScaffold(
      scaffoldBackgroundColor: Colors.orange,
      bodyBackgroundColor: Colors.orange,
      body: homeBody(),
      toolbarLeftIcon: Icons.arrow_back,
      toolbarLeftIconType: BaseToolbar.IMAGE_TYPE_ICON,
      toolbarTitle: 'Hello',
      toolbarRightText: 'Done',
      toolbarHeight: 80,
      toolbarTitleTextStyle: toolbarTitleTextStyle(),
      isScreenLoading: _isScreenLoading,
      isScreenLoadingWithBackground: _showLoading,
      screenLoaderView: emptyLoaderView(),
    );
  }

  ClipRRect emptyLoaderView() {
    return ClipRRect(
      borderRadius: BorderRadius.circular(5),
      child: Container(
        height: 50,
        width: 100,
        color: Colors.white,
        child: Center(
          child: Text(
            'Loading...',
          ),
        ),
      ),
    );
  }

  TextStyle toolbarTitleTextStyle() {
    return TextStyle(
        color: Colors.orange, fontSize: 22, fontWeight: FontWeight.bold);
  }

  Container homeBody() {
    return Container(
      color: Colors.green,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.all(50),
            child: RaisedButton(
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(5)),
              child: Text('Empty screen loading...'),
              onPressed: () {
                setState(() {
                  _isScreenLoading = true;
                });
                Future.delayed(Duration(seconds: 3), () {
                  setState(() {
                    _isScreenLoading = false;
                  });
                });
              },
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(50),
            child: RaisedButton(
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(5)),
              child: Text('Show Loading...'),
              onPressed: () {
                setState(() {
                  _showLoading = true;
                });
                Future.delayed(Duration(seconds: 3), () {
                  setState(() {
                    _showLoading = false;
                  });
                });
              },
            ),
          ),
        ],
      ),
    );
  }
}


lib/base_toolbar.dart
import 'package:flutter/material.dart';

class BaseToolbar extends StatefulWidget {
  static const String IMAGE_TYPE_IMAGE = "image";
  static const String IMAGE_TYPE_ICON = "icon";

  // Toolbar
  /// Set icon on left side of the toolbar.
  /// Icon can be an [String] value if you use an assets
  ///                   OR
  /// Icon can be an [IconData] if you use Icons.
  final leftIcon;

  /// If you set left icon you need to provide the type of icon.
  /// Use this to set type of icon can be from below two option.
  /// * [BaseToolbar.IMAGE_TYPE_IMAGE] if icon is from assets.
  /// * [BaseToolbar.IMAGE_TYPE_ICON] if icon is type of Icons.
  ///
  /// @Note: If you not provide any type to left icon then left icon will
  /// not display in toolbar.
  final String? leftIconType;

  /// Set left icon click event if you have used [leftIcon] &
  /// [leftIconType].
  final Function? leftIconClick;

  /// Set icon on right side of the toolbar.
  /// Icon can be an [String] value if you use an assets
  ///                   OR
  /// Icon can be an [IconData] if you use Icons.
  final rightIcon;

  /// If you set right icon you need to provide the type of icon.
  /// Use this to set type of icon can be from below two option.
  /// * [BaseToolbar.IMAGE_TYPE_IMAGE] if icon is from assets.
  /// * [BaseToolbar.IMAGE_TYPE_ICON] if icon is type of Icons.
  ///
  /// @Note: If you not provide any type to right icon then right icon will
  /// not display in toolbar.
  final String? rightIconType;

  /// Set right icon click event if you have used [rightIcon] &
  /// [rightIconType].
  final Function? rightIconClick;

  /// Set title on toolbar.
  final String? title;

  /// Set text at right side of toolbar.
  final String? rightText;

  /// Set click event of right side text.
  /// @Note: It will not work if [rightText] is empty or not set.
  final Function? rightTextClick;

  /// Set text at left side of toolbar.
  final String? leftText;

  /// Set click event of left side text.
  /// @Note: It will not work if [leftText] is empty or not set.
  final Function? leftTextClick;

  /// Manage height of toolbar.
  /// By default it height is [60] and it can not be less than [50].
  final double toolbarHeight;

  /// Set custom text style of left text in toolbar.
  final TextStyle? leftTextStyle;

  /// Set custom text style of title text in toolbar.
  final TextStyle? titleTextStyle;

  /// Set custom text style of right text in toolbar.
  final TextStyle? rightTextStyle;

  /// Use to set only color of left icon.
  final Color? leftIconColor;

  /// Use to set only color of right icon.
  final Color? rightIconColor;

  /// Make custom widget to set at left side of toolbar.
  final Widget? leftChild;

  /// Make custom widget to set at center of toolbar.
  final Widget? centerChild;

  /// Make custom widget to set at right side of toolbar.
  final Widget? rightChild;

  /// Set toolbar background color.
  final Color backgroundColor;

  BaseToolbar({
    this.leftIconType,
    this.leftIcon,
    this.leftIconColor,
    this.leftIconClick,
    this.rightIcon,
    this.rightIconColor,
    this.rightIconType,
    this.rightIconClick,
    this.title,
    this.rightText,
    this.rightTextClick,
    this.leftText,
    this.leftTextClick,
    this.backgroundColor = Colors.white,
    this.toolbarHeight = 60,
    this.leftTextStyle,
    this.titleTextStyle,
    this.rightTextStyle,
    this.leftChild,
    this.centerChild,
    this.rightChild,
  })  : assert(toolbarHeight > 50),
        assert(
            leftIcon != null
                ? leftIcon is String || leftIcon is IconData
                : true,
            "Please provide icon type in string(assets or local path) or IconData."),
        assert(
            rightIcon != null
                ? rightIcon is String || rightIcon is IconData
                : true,
            "Please provide icon type in string(assets or local path) or IconData.");

  @override
  _BaseToolbarState createState() => _BaseToolbarState();
}

class _BaseToolbarState extends State<BaseToolbar> {
  ThemeData? themeData;

  @override
  Widget build(BuildContext context) {
    themeData = Theme.of(context);
    return Container(
      color: widget.backgroundColor,
      height: widget.toolbarHeight,
      child: Row(
        children: <Widget>[
          Expanded(
              flex: 1,
              child: (widget.leftChild != null
                  ? widget.leftChild
                  : widget.leftIcon != null
                      ? leftIconWidget(widget.leftIconClick)
                      : leftTextTitle(widget.leftText!, widget.leftTextClick))!),
          Expanded(
              flex: 4,
              child: (widget.centerChild != null
                  ? widget.centerChild
                  : centerText(widget.title!))!),
          Expanded(
              flex: 1,
              child: (widget.rightChild != null
                  ? widget.rightChild
                  : widget.rightIcon != null
                      ? rightIconWidget(widget.rightIconClick)
                      : rightTextTitle(
                          widget.rightText!, widget.rightTextClick))!),
        ],
      ),
    );
  }

  Widget leftIconWidget(leftIconClick) {
    return GestureDetector(
      onTap: leftIconClick,
      child: Padding(
        padding: EdgeInsets.all(15),
        child: getLeftIcon(
            widget.leftIconType, widget.leftIcon, widget.leftIconColor),
      ),
    );
  }

  Widget getLeftIcon(leftIconType, leftIcon, leftIconColor) {
    if (leftIconType == BaseToolbar.IMAGE_TYPE_IMAGE) {
      return Image.asset(
        leftIcon,
        height: 30,
        width: 30,
      );
    } else if (leftIconType == BaseToolbar.IMAGE_TYPE_ICON) {
      return Icon(
        leftIcon,
        color: leftIconColor ?? themeData!.primaryColor,
      );
    } else {
      return Container();
    }
  }

  Widget leftTextTitle(String leftText, leftTextClick) {
    return Container(
      child: leftText == null || leftText.isEmpty
          ? Container()
          : GestureDetector(
              onTap: leftTextClick,
              child: Text(
                leftText,
                style: widget.leftTextStyle ??
                    themeData!.textTheme.subtitle2!
                        .copyWith(color: themeData!.primaryColor),
                textAlign: TextAlign.center,
              ),
            ),
    );
  }

  Widget centerText(String title) {
    return Container(
      child: title == null || title.isEmpty
          ? Container()
          : Text(
              title,
              style: widget.titleTextStyle ??
                  themeData!.textTheme.headline6!.copyWith(
                      color: Colors.black, fontWeight: FontWeight.w500),
              textAlign: TextAlign.center,
            ),
    );
  }

  Widget rightTextTitle(String rightText, rightTextClick) {
    return Container(
      child: rightText == null || rightText.isEmpty
          ? Container()
          : GestureDetector(
              onTap: rightTextClick,
              child: Text(
                rightText,
                style: widget.rightTextStyle ??
                    themeData!.textTheme.subtitle2!
                        .copyWith(color: themeData!.primaryColor),
                textAlign: TextAlign.center,
              ),
            ),
    );
  }

  Widget rightIconWidget(rightIconClick) {
    return GestureDetector(
      onTap: rightIconClick,
      child: Padding(
        padding: EdgeInsets.all(15),
        child: getRightIcon(
            widget.rightIconType, widget.rightIcon, widget.rightIconColor),
      ),
    );
  }

  Widget getRightIcon(rightIconType, rightIcon, rightIconColor) {
    if (rightIconType == BaseToolbar.IMAGE_TYPE_IMAGE) {
      return Image.asset(
        rightIcon,
        height: 30,
        width: 30,
      );
    } else if (rightIconType == BaseToolbar.IMAGE_TYPE_ICON) {
      return Icon(
        rightIcon,
        color: rightIconColor ?? themeData!.primaryColor,
      );
    } else {
      return Container();
    }
  }
}
